rep-0145.rst

```diff
@@ -0,0 +1,190 @@
+REP: 145
+Title: Conventions for IMU Sensor Drivers
+Author: Paul Bovbel <pbovbel@clearpathrobotics.com>
+Status: Draft
+Type: Informational
+Content-Type: text/x-rst
+Created: 02-Feb-2015
+Post-History: 02-Feb-2015
+
+
+Abstract
+========
+
+This REP defines common parameters, topics, namespaces, and data processing conventions for drivers of Inertial Measurement Unit (IMU) sensors. This includes accelerometers, gyroscopes, magnetometers, and any combination thereof.
+
+Motivation
+==========
+
+This REP seeks to standardize an interface for IMU drivers in the ROS ecosystem, providing for IMU-type sensors with a variety of capabilities and conventions. By formalizing a common approach based on existing driver implementations, this REP serves as a guideline for new driver development, while minimizing the changes required to non-conforming implementations. This REP discourages the in-driver manipulation of sensor data, by specifying the conditions where sensor data will be properly handled by ROS transform and coordinate specifications. The REP also specifies fallback approaches to correcting sensor data that does not conform to ROS conventions.
+
+Specification
+=============
+
+Frame Conventions
+-----------------
+
+An IMU device may measure data with respect to two frames, specified by the manufacturer:
+
+* The device **body frame** represents the device's internal reference frame (the "input reference axes", IRA). It may be found in the device specification documents, and is sometimes indicated directly on the package of the device. This frame is fixed to the device (i.e. it is a moving coordinate frame).
+
+* If an orientation estimate is provided (see `Data Reporting`_), the device **world frame** represents the external reference with respect to which the **body frame** is indicated. Two out of three degrees of freedom of the **body frame** w.r.t. the **world frame** are usually considered "fixed" through gravity, with the relevant conventions from REP 103 [1]_ :
+
+  - For NED type IMUs, the orientation of the world frame is x-north, y-east, z-down, relative to magnetic north.
+
+  - For ENU type IMUs, the orientation of the world frame is x-east, y-north, z-up, relative to magnetic north.
+
+* If the device does not have an absolute yaw reference (magnetometer), the world frame will only be aligned in roll and pitch (due to gravity), while the orientation around the z axis can be arbitrary.
+
+* If the device does not have an absolute gravity reference (accelerometer), the world frame is not aligned to any external reference and aligned entirely to the power-on orientation of the sensor.
```

Comment:
I'm having second thoughts about this bullet point, and now I think it should be removed (although I know I put it in there). To recap why I put it in: On our old robots, we only had a single gyro that measured the yaw axis, so we put only that into the orientation, and it worked with `robot_pose_ekf`. But what we have actually been representing there is *not* an arbitrary orientation; instead, we've been faking a gravity-aligned orientation.

@paulbovbel : Please wait a bit before making this change in case somebody else has a different opinion.

Reply:
Agreed, removing

Reply:
It would be useful.. in space!

Agreed, removing.

Reply:
I'm ambivalent about this. I think my thinking here was to describe what _if anything_ should be written to `orientation` if the IMU has neither an accelerometer nor a magnetometer.